### PR TITLE
Update Mod and RandBetween type handling

### PR DESCRIPTION
**Pull request by gregli-msft**
From a type perspective, Mod should act like / (division).  The arguments should be symmetric with a Decimal result only if both arguments are Decimal.  This is not the case, there is a bug here where the first argument is dictating the result of the function alone.  The same holds for RandBetween which has the same problem.  This PR fixes both functions and adds many tests.  https://github.com/microsoft/Power-Fx/issues/2356

Here, the second result should be 1 with no decimal as the calculation should be done in Float.
```
>> Mod( 3.0000000000000000000003, 2.0000000000000000000002 )
1.0000000000000000000001

>> Mod( 3.0000000000000000000003, Float(2.0000000000000000000002) )
1.0000000000000000000003

>> Mod( Float(3.0000000000000000000003), 2.0000000000000000000002 )
1

>> Mod( Float(3.0000000000000000000003), Float(2.0000000000000000000002) )
1
```

Likewise, RandBetween has the wrong answer for the second case, where the second argument is coerced to a decimal (while losing digits).
```
>> RandBetween( 12345678912345678912345678, 12345678912345678912345678 )
12345678912345678912345678

>> RandBetween( 12345678912345678912345678, Float(12345678912345678912345678) )
12345678912345697471862876

>> RandBetween( Float(12345678912345678912345678), 12345678912345678912345678 )
1.2345678912345679E+25

>> RandBetween( Float(12345678912345678912345678), Float(12345678912345678912345678) )
1.2345678912345679E+25
```

In the process, it was discovered that RandBetween will overflow if more than half of the allowable range is input.  This has also been fixed.  https://github.com/microsoft/Power-Fx/issues/2357

``` 
>> RandBetween( -79228162514264337593543950330, 79228162514264337593543950330 )
Value was either too large or too small for a Decimal.

>> RandBetween( Float("-1e308"), Float("1e308") )
<Error: The function RandBetween returned a non-finite number.>
```

And finally, one more RandBetween bug https://github.com/microsoft/Power-Fx/issues/2358, second example here:
```
>> RandBetween( 79228162514264337593543950330, 79228162514264337593543950332 )
79228162514264337593543950330

>> RandBetween( 79228162514264337593543950330, 79228162514264337593543950332 )
79228162514264337593543950333

>> RandBetween( 79228162514264337593543950330, 79228162514264337593543950332 )
79228162514264337593543950331
```

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Rand.txt [hunk 1]
```diff
@@ -59,4 +59,19 @@ Error({Kind:ErrorKind.Numeric})
...
\ No newline at end of file
+Blank()
+
+>> RandBetween( 9007199254740992 /* MaxFloatInt */, 9007199254740992 /* MaxFloatInt */ )
+9007199254740992
+
+>> RandBetween( -9007199254740992 /* MaxFloatInt */, -9007199254740992 /* MaxFloatInt */ )
+-9007199254740992
+
+>> Sum(Sequence(100), With( { r: RandBetween( 9007199254740985, 9007199254740992 /* MaxFloatInt */ ) }, If( r >= 9007199254740985 And r <= 9007199254740992, 1, Error( "Result out of range" ) ) ) )
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Rand.txt:70]: >r >= 9007199254740985 And r <= 9007199254740992 [](http://example.com/codeflow?start=108&length=47)

Suggestion: `9007199254740985 <= r And r <= 9007199254740992`: easier to understand the boundaries

**Comment by CarlosFigueiraMSFT** [comment]
> // this is the one result in this file that should be decimal, everything else is done in floating point

Not your changes, but this comment is incorrect, as the test on line 60 is also a decimal result

---
Refers to: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/OpMatrix_Div_Float_DecimalSupport.txt:56 in f8bfd92. [](commit_id = f8bfd9268e95cd7a8d5c0aeb37ff3a1ae288e78c, deletion_comment = False)

File: src/tests/Microsoft.PowerFx.Core.Tests/Microsoft.PowerFx.Core.Tests.csproj [hunk 2]
```diff
@@ -49,6 +49,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <None Remove="expressiontestcases\ModT_Decimal.txt" />
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/Microsoft.PowerFx.Core.Tests.csproj:52]: ?!?

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ModT.txt [hunk 3]
```diff
@@ -5,3 +5,33 @@ Table({Value:0},{Value:1})
 
 >> Mod(3, [-1, 0, 1, 2, 3, Blank()])
 Table({Value:0},{Value:Error({Kind:ErrorKind.Div0})},{Value:0},{Value:1},{Value:0},{Value:Error({Kind:ErrorKind.Div0})})
+
+>> Mod([1,2,3,4,5,6,7,8,9,10],3)
```
Comment (LucGenetier) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ModT.txt:9]: Would add a couple of tests with negative numbers

**Comment by CarlosFigueiraMSFT** [comment]
> #SETUP: ConsistentOneColumnTableResult

Please also add PowerFxV1CompatibilityRules; before V1 we wouldn't have ErrorKind.NotApplicable with tables of different sizes

---
Refers to: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ModT.txt:1 in f8bfd92. [](commit_id = f8bfd9268e95cd7a8d5c0aeb37ff3a1ae288e78c, deletion_comment = False)

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Rand.txt (hunk 1 above)
Comment (gregli-msft) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Rand.txt]: Done.

File: src/tests/Microsoft.PowerFx.Core.Tests/Microsoft.PowerFx.Core.Tests.csproj (hunk 2 above)
Comment (gregli-msft) [reply src/tests/Microsoft.PowerFx.Core.Tests/Microsoft.PowerFx.Core.Tests.csproj]: Oops, fixed.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Mod_Decimal.txt [hunk 4]
```diff
@@ -16,4 +16,59 @@
...
 >> Mod(79228162514264337593543950335, 123)
-99
\ No newline at end of file
+99
+
+>> Mod( 79228162514264337593543950335, 79228162514264337593543950335 )
+0
+
+>> Mod( 79228162514264337593543950335, 79228162514264337593543950334 )
+1
```
[the remark this answers is not in the document]
Comment (gregli-msft) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Mod_Decimal.txt:25]: Done.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ModT.txt (hunk 3 above)
Comment (gregli-msft) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ModT.txt:9]: Done.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by CarlosFigueiraMSFT — APPROVED** [review]
:shipit:
